torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -83,7 +83,7 @@ struct KernelArgumentHolder {
       const at::Tensor& val,
       c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
     changed = true;
-    ExtractSizeStride ess(val, std::move(broadcasted_size));
+    ExtractSizeStride ess(val, broadcasted_size);
```

Comment:
why not move?

Reply:
It doesn't buy us anything in this case. But more importantly, std::move() is a code smell - if the types are well designed you should rarely need to use it in general purpose code, and there are cases where's it's actively harmful.

(yes, I noticed a common pattern in pytorch where parameters are passed by value then moved - I don't think it's a great idiom or something to emulate - passing by const ref is a much simpler alternative so there's no excuse of it).

Reply:
Fair enough, I didn't have such a strong opinion on std::move, but I believe we primarily use refs in the rest of the code base, so see no harm in converting this as well to a ref. @jjsjann123

Reply:
`c10::optional` makes move much safer (easier to catch expired content) so I usually don't mind it much and do so for slight better perf.
Wouldn't mind `const ref` either.

Nitpick: if we are getting rid of the `move` here, maybe we want to change the constructor of `ExtractSizeStride` to use const ref just so we don't have the additional copy.

Reply:
I'm curious, how does c10::optional help catch expired content? (moving an optional just moves the contained value, it does not make source "empty" - or did I misunderstand the comment?)

Also, what is the perf benefit here? Moving an ArrayRef seems to have exactly the same cost as copying an ArrayRef, right?

One of the reason I'm suspicious of std::move() here is exactly the fact that it suggests that something special is going on: a noticeable performance optimization and/or move-only semantics. If overused, this is becoming problematic (you can no longer trust it and reason about what's going on, never mind the additional verbosity)

Reply:
Yes you understand the comment correctly. I was wrong when I expect the move to clear the init of the source (making it empty), but it does not :cry: .

In the context of ArrayRef, I agree that moving / copying would be identical. I should work on my programming habit and try to avoid unnecessary wordiness :laughing: